src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs

```diff
@@ -133,18 +134,16 @@ public bool TryRead(out T value)
         [MethodImpl(MethodImplOptions.NoInlining)]
         private void GetNextSpan()
         {
-            if (!Sequence.IsSingleSegment)
+            // If Sequence is a single segment, TryGet will return false
+            SequencePosition previousNextPosition = _nextPosition;
+            while (Sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<T> memory, advance: true))
             {
-                SequencePosition previousNextPosition = _nextPosition;
-                while (Sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<T> memory, advance: true))
+                if (memory.Length > 0)
                 {
-                    _currentPosition = previousNextPosition;
                     CurrentSpan = memory.Span;
+                    _currentPosition = previousNextPosition;
```

Comment:
This will break. We should have an empty span when we get to the end. Current position won't get updated either. Can you add tests to ensure that behavior does not change in the end state? Empty sequences at the end are particularly concerning.